### ISSUE TITLE
`typeof SomeClass` is "function" in TS/JS but "object" in TSTL

### ISSUE THREAD

**Issue by probable-basilisk** [TypeScriptToLua/TypeScriptToLua#1207]
```ts
class SomeClass{}
console.log(typeof SomeClass)
```
produces "object" is TSTL but "function" in TS [(playground)](https://typescripttolua.github.io/play#code/MYGwhgzhAEAqAWBLAdgc2gbwL4CgfAHtkICQBTAOhANQAoAiAdXjABdpEYEVUByGVgE8ADmQBc9ADTQhoggDM4SNAEocQA). 

Since Typescript expects "function" for this, this creates problems when trying to narrow down types for functions that can take class constructors (e.g., JSX createElement).

**Comment by GlassBricks** [comment]
I think this inconsistency is unavoidable given how TS classes are implemented (functions cannot have properties in lua).

For a workaround, here is one way to detect TSTL classes that should work for most cases:
```ts
function isClass(obj: unknown) {
    return type(obj) === "table" && rawget(obj as any, "prototype")?.constructor === obj
}
```

**Comment by probable-basilisk** [comment]
Yes, like the difference between Lua and Javascript notions of truthiness, I'm perfectly happy to declare that actually Javascript is wrong. But it would be nice to be able to tell the TS type checker this, without workarounds that require escaping into Lua (although you can write type-predicated functions like
```ts
function isActualFunction(f: unknown): f is ((...args: any[]) => any) {
  return luatype(f) == "function"
}

function callOrConstruct(f: (() => Elem) | typeof Elem): Elem {
  if (isActualFunction(f)) {
    return f()
  } else {
    return new f()
  }
}
```
which will satisfy the type checker and hide a lot of the ugliness).
